### PR TITLE
Add ACPI TPM2 table generator

### PR DESCRIPTION
**Pull request by nvidia-dmach**
# Description
- Add AcpiTpm2LibArm for generating ACPI TPM2 table using the information obtained from Tpm2InterfaceInfo CM object.
- Add helper function HexDump for printing hex dump of CM Object fields. Also merge multiple flavors of PrintCharX into one function PrintChars by using the field length.
- Add parser for Tpm2 CM object, which uses HexDump for dumping the 12 bytes of "StartMethodParameters" field.

- [ ] Breaking change?
- [ ] Impacts security?
- [ ] Includes tests?

## How This Was Tested
- Verified that TPM2 table was created properly on platform that produces Tpm2InterfaceInfo CM object.
- Called ParseCmObjDesc() and ensured that all fields of the CM objects were printed correctly.

## Integration Instructions

N/A

### CODE REVIEW

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (pierregondois) [inline DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:43]: I might be missing information because I didn't find the spec defining the revision 3 of the table, but either:
- This struct should be added to MdePkg/Include/IndustryStandard/Tpm2Acpi.h if it is really different from the revision 3
- Or the StartMethodParameters/Laml/Lasa fields are commented because of the fact that StartMethodParameters is of variable length. In such case you should add the following struct  (to Tpm2Acpi.h) to access them:

typedef struct {
  UINT32                      Laml;                          // Optional
  UINT64                      Lasa;                          // Optional
} SOME_NAME;

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (pierregondois) [inline DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:146]: I didn't find the spec defining the revision 3 of the table,
would you have a link by any chance ?

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (pierregondois) [inline DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:151]: Cf: TCG Hardware Interface Description Table Format for TPM 2.0

```
The length of this table starting from the Signature field up to
and including the platform specific parameters field. (52 +
size of platform specific parameters)
```

Apparently it is possible to infer the number of bytes used based in the "Start Method Specific Parameters" based on the value of "Start Method", but I didn't find how to do that exactly:
```
The content of the Start Method specific parameters is
determined by the Start Method used by the system’s TPM
device interface. This field contains values that may be used
to initiate command processing.
If the Start Method value is 2, then this field is at least four
bytes in size and the first four bytes must be all zero.
If the Start Method value is 11 then this field is 12 bytes in
size and is described in Table 9.
If LAML and LASA are present, then this field is 12 bytes in
size.
```

As the number of bytes used seems to be variable for one value of the Start Method, I think you should add another field in CM_ARM_TPM2_INTERFACE_INFO to specify how many bytes of StartMethodParameters[12] will be used.

**Comment by pierregondois** [comment]
Also:
- I saw there is a revision 5 of the ACPI table, do you plan to add support for it (I didn't check how much modification this represent)
- Do you plan to add a parser for the table in acpiview ?
- There seems to be an SSDT table corresponding to this ACPI table, do you plan to generate it aswell ?
Regards,
Pierre

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:151]: Good point. At the first thought, I guess we would need to add a size field for StartMethodParameters[]. But let me think about it a little more.

File: DynamicTablesPkg/Include/ArmNameSpaceObjects.h
Comment (nvidia-dmach) [reply DynamicTablesPkg/Include/ArmNameSpaceObjects.h:1373]: Will do

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:43]: Actually, this was meant to be a local instantiation of the one defined in Tpm2Acpi.h, so that we could access Laml and Lasa directly. The generic one in Tpm2Acpi.h already has the fields defined for V4 but commented out.

I will spend some time to find a better implementation for this.

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:146]: I thought I saw a reference to rev 3 somewhere but I can't find it anymore. I will change it to "if (AcpiTableInfo->AcpiTableRevision < EFI_TPM2_ACPI_TABLE_REVISION_4)" instead, following what had been done in Tcg2Acpi.c:
https://github.com/tianocore/edk2/blob/3b2025969e6e8a2f6542996182cd4132868641c6/SecurityPkg/Tcg/Tcg2Acpi/Tcg2Acpi.c#L824.

File: DynamicTablesPkg/Library/Common/TableHelperLib/ConfigurationManagerObjectParser.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Common/TableHelperLib/ConfigurationManagerObjectParser.c:685]: Will do

**Comment by nvidia-dmach** [comment]
Hi Pierre,
Thanks for your review. Below are the answers to your questions:
> * I saw there is a revision 5 of the ACPI table, do you plan to add support for it (I didn't check how much modification this represent)
No, not at the moment. We can add rev 5 in the future patch.
> * Do you plan to add a parser for the table in acpiview ?
I suppose you mean the CM parser. If so, yes, it was added as the 3rd patch of this PR.
> * There seems to be an SSDT table corresponding to this ACPI table, do you plan to generate it aswell ?
The SSDT one is platform specific. So I think it's better to do in platform code instead.

BTW, I will be OOTO next week. Please expect some delay in my response.

Thanks,
Dat

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (pierregondois) [reply DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:43]: As it is unknown how long the `StartMethodParameters` field will be in the struct, I don't think this is possible to define a struct to access the Laml/Lasa fields. You will have to typecast a pointer to the end of the `StartMethodParameters` field to the type `SOME_NAME` (cf. above)

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (pierregondois) [reply DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:146]: The issue is that without the spec, it's hard to check the table generation is valid.
If we're not able to find the spec., it might be better to limit the generation of the table to the Revision 4 onward (and don't support the version 3).

**Comment by pierregondois** [comment]
> Hi Pierre, Thanks for your review. Below are the answers to your questions:
> 
> > * I saw there is a revision 5 of the ACPI table, do you plan to add support for it (I didn't check how much modification this represent)
> >   No, not at the moment. We can add rev 5 in the future patch.
> > * Do you plan to add a parser for the table in acpiview ?
> >   I suppose you mean the CM parser. If so, yes, it was added as the 3rd patch of this PR.

There is a module in another package which allows to display ACPI tables, for instance:
ShellPkg/Library/UefiShellAcpiViewCommandLib/Parsers/Srat/

I was wondering if you were planning to add support to display this table aswell

> > * There seems to be an SSDT table corresponding to this ACPI table, do you plan to generate it aswell ?
> >   The SSDT one is platform specific. So I think it's better to do in platform code instead.

I think that all the ACPI tables are platform specific.
The SSDT table seems complicated to generate, I was just wondering, no worries if we don't generate it.

> 
> BTW, I will be OOTO next week. Please expect some delay in my response.
> 
> Thanks, Dat

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:43]: Actually, according to the spec, "If LAML and LASA are present, then this field is 12 bytes in size." Hence, the offsets to Laml/Lasa are fixed.

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:146]: Makes sense. I will remove version 3 support then.

**Comment by nvidia-dmach** [comment]
Hi Pierre, Sami,

I have made fixes to my patch set. Please review again when you get a chance.

Thanks,
Dat

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:10]: Yes, will do.

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:49]: I see. I will find a better way to implement this.

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:156]: It sounds like when Start Method value is 2, the parameter field can still have variable size as long as it is at least 4. Because of that we can't use GetStartMethodLength(), but have to rely on the Size field from CM object. But I agree with you that we can add sanity checks to ensure the size as expected based on Start Method value.

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:155]: Will do

File: DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c:46]: Agree. I will remove this structure.

File: MdePkg/Include/IndustryStandard/Tpm2Acpi.h
Comment (nvidia-dmach) [reply MdePkg/Include/IndustryStandard/Tpm2Acpi.h:22]: Will do

File: DynamicTablesPkg/Library/Common/TableHelperLib/ConfigurationManagerObjectParser.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Common/TableHelperLib/ConfigurationManagerObjectParser.c:686]: Will add StartMethodParametersSize field in my next patch.

File: DynamicTablesPkg/Include/ArmNameSpaceObjects.h
Comment (nvidia-dmach) [reply DynamicTablesPkg/Include/ArmNameSpaceObjects.h:1378]: Ok, I will keep StartMethodParametersSize since the spec doesn't clearly state the fixed size for when Start Method value is 2.

**Comment by nvidia-dmach** [comment]
Hi Pierre, Sami,

Thank you for your feedbacks. I have updated my patches to address them. Please review again when you get a chance.

Thanks,
Dat

File: DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c
Comment (pierregondois) [inline DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c:72]: The generator should only support the revision 4 right ?
So I think this parameter should not be needed

**Review by pierregondois — COMMENTED** [review]
Hello Dat,
I added some comments. I think that with the following changes (if you think they are correct), the patch would be ok for me,

[patch.txt](https://github.com/user-attachments/files/16428393/patch.txt)

Regards,
Pierre

File: DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c:213]: Good catch! I reworded it but forgot to swap the arguments.

File: DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c:73]: I just wanted to make it scalable for future rev 5 support (max size 16). The CM object creators only need to fill in the parameter size that they actually use. For example, 4 bytes for StartMethod = 2, then the generator would know which rev and make the right shift for Laml/Lasa fields. So the CM object creators only supply the data and don't need to worry about the actual table format.

File: DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c:259]: Will remove this check

File: DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c:72]: This is to prepare for rev 5. So I want to keep it if it's ok.

File: DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c
Comment (pierregondois) [reply DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c:73]: Ok understood,
would it be possible to add a comment about this in:
`DynamicTablesPkg/Include/ArchCommonNameSpaceObjects.h`
to inform that:
- for the `Lasa` field: If `Lasa` is populated, then `Laml` should be populated (and similar comment for the `Laml field`)
- for the `StartMethodParametersSize` field, if `Laml` and `Lasa` fields are populated, the field can be left empty

File: DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c
Comment (pierregondois) [reply DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c:72]: Ok, sorry I forgot about the Rev 5.
IIUC, then the only changes required should be:
[patch.txt](https://github.com/user-attachments/files/16438416/patch.txt)

File: DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c:73]: Actually, thinking about this a little more, I found some issues with it. So I am making changes based on your first patch.txt. I will resubmit once my testing is done. Thank you for your review.

File: DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c
Comment (nvidia-dmach) [reply DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c]: I removed the revision check for now. It can be added later when we support revision 5.

File: DynamicTablesPkg/Include/ArchCommonNameSpaceObjects.h
Comment (samimujawar) [inline DynamicTablesPkg/Include/ArchCommonNameSpaceObjects.h:682]: Minor: Is it possible to place the StartMethodParametersSize field before the StartMethodParameters field, please?
The reason I say this is in the future we may extend the ConfigurationManager Object parser to extract the StartMethodParametersSize value and then use it to print the StartMethodParameters. Something similar to what we do today in Acpiview.

Other than that this patch series looks good to me.

Comment (nvidia-dmach) [reply DynamicTablesPkg/Include/ArchCommonNameSpaceObjects.h]: I have swapped the Size field. Thanks.

**Comment by lgao4** [comment]
The change in MdePkg is good to me.

**Review by samimujawar — APPROVED** [review]

**Comment by samimujawar** [comment]
Hi Liming @lgao4, Mike @mdkinney,

This is a new feature to add TPM2 table generation capability in Dynamic Tables Framework.

This patch set had almost completed the review before 5th Aug. Would it be possible to apply the push label so that it can catch this release cycle, please?

Regards,

Sami Mujawar

**Comment by samimujawar** [comment]
@nvidia-dmach I think you may need to rebase and push this series again. Can you take a look, please?

**Comment by nvidia-dmach** [comment]
Hi Sami,
I just rebased my patches. Please review again. Thanks.

**Comment by lgao4** [comment]
> Hi Liming @lgao4, Mike @mdkinney,
> 
> This is a new feature to add TPM2 table generation capability in Dynamic Tables Framework.
> 
> This patch set had almost completed the review before 5th Aug. Would it be possible to apply the push label so that it can catch this release cycle, please?
> 
> Regards,
> 
> Sami Mujawar

Yes. But, this pach needs approve from the maintainer or reviewer of MdePkg and DynamicTablesPkg. I will give my approval. After DynamicTablesPkg maintainer gives the approval, I will add push label for this change.

**Review by lgao4 — APPROVED** [review]

**Review by samimujawar — APPROVED** [review]

**Comment by nvidia-dmach** [comment]
Hi Sami, Liming,

Thanks for the approvals. Is there anything I need to do for this patch set to merge?

Thanks,
Dat

**Pull request merged**
